src/Microsoft.AspNetCore.Dispatcher/DispatcherValueEndpointSelector.cs

```diff
@@ -2,54 +2,244 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.AspNetCore.Dispatcher.Internal;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
 
 namespace Microsoft.AspNetCore.Dispatcher
 {
     public class DispatcherValueEndpointSelector : EndpointSelector
     {
+        private static readonly IReadOnlyList<SimpleEndpoint> EmptyEndpoints = Array.Empty<SimpleEndpoint>();
+
+        private EndpointSelectorContext _context;
+        private Cache _cache;
+
+        private ILogger<DispatcherValueEndpointSelector> _logger;
+        private bool _servicesInitialized;
+
+        public override void Initialize(IEndpointCollectionProvider endpointCollectionProvider)
+        {
+            var cache = new Cache(endpointCollectionProvider.Endpoints);
+            Volatile.Write(ref _cache, cache);
+        }
+
         public override Task SelectAsync(EndpointSelectorContext context)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var dispatcherFeature = context.HttpContext.Features.Get<IDispatcherFeature>();
+            _context = context;
+            if (!_servicesInitialized)
+            {
+                EnsureServices(context.HttpContext);
+            }
 
-            for (var i = context.Endpoints.Count - 1; i >= 0; i--)
+            var dispatcherFeature = _context.HttpContext.Features.Get<IDispatcherFeature>();
+            var candidates = SelectCandidates(dispatcherFeature.Values);
+            _context.Endpoints.Clear();
+            switch (candidates.Count)
             {
-                var endpoint = context.Endpoints[i] as IDispatcherValueSelectableEndpoint;
-                if (!CompareRouteValues(dispatcherFeature.Values, endpoint.Values))
+                case 0:
+                    _logger.NoEndpointsMatched(dispatcherFeature.Values);
+                    break;
+
+                case 1:
+                    _logger.EndpointMatched(candidates[0].DisplayName);
+                    break;
+
+                default:
+                    var endpointNames = string.Join(
+                            Environment.NewLine,
+                            candidates.Select(a => a.DisplayName));
+                    _logger.AmbiguousEndpoints(endpointNames);
+                    break;
+            }
```

Comment:
I'm not sure how this code made it into here.. There should be one block like this and it should be in `DispatcherBase`.

These policies run in a sequence, so it's not ambiguous if we have multiple endpoints - the ambiguity could be resolved by a subsequent selector